_posts/2018-05-10-Rust-1.26.md

```diff
@@ -0,0 +1,545 @@
+fn foo() -> impl Trait {
+    5
+}
+```
+
+This doesn't create a trait object, it's like we had written `-> i32`, but
+instead, we're only mentioning the part about `Trait`. We get static
+dispatch, but we can hide the real type like this.
+
+Why is this useful? One good use is closures. Remember that closures in
+Rust all have a unique, un-writable type, yet implement the `Fn` trait.
+This means that if your function returns a closure, you can do this:
+
+```rust
+// before
+fn foo() -> Box<Fn(i32) -> i32> {
+    Box::new(|x| x + 1)
+}
+
+// after
+fn foo() -> impl Fn(i32) -> i32 {
+    |x| x + 1
+}
+```
+
+No boxing, no dynamic dispatch. A related scenario happens when returning
+iterators. Not only do iterators often include closures, but since they
+nest, you get quite deeply nested types. For example:
+
+```rust
+fn foo() {
+    vec![1, 2, 3]
+        .into_iter()
+        .map(|x| x + 1)
+        .filter(|x| x % 2 == 0)
+}
+```
+
+when compiled, gives this error:
+
+```text
+error[E0308]: mismatched types
+ --> src/main.rs:5:5
+  |
+5 | /     vec![1, 2, 3]
+6 | |         .into_iter()
+7 | |         .map(|x| x + 1)
+8 | |         .filter(|x| x % 2 == 0)
+  | |_______________________________^ expected (), found struct `std::iter::Filter`
+  |
+  = note: expected type `()`
+             found type `std::iter::Filter<std::iter::Map<std::vec::IntoIter<{integer}>, [closure@src/main.rs:7:14: 7:23]>, [closure@src/main.rs:8:17: 8:31]>`
+```
+
+That's a huge 'found type'. Each adapter in the chain adds a new type.
+Additionally, we have that closure in there. Previously, we'd have had
+to use a trait object here, but now we can simply do
+
+```rust
+fn foo() -> impl Iterator<Item = i32> {
+    vec![1, 2, 3]
+        .into_iter()
+        .map(|x| x + 1)
+        .filter(|x| x % 2 == 0)
+}
+```
+
+and be done with it. Working with futures is very similar.
+
+It's important to note that sometimes trait objects are still
+what you need. You can only use `impl Trait` if your function returns
+a single type; if you want to return multiple, you need dynamic dispatch.
+For example:
+
+```rust
+fn foo(x: i32) -> Box<Iterator<Item = i32>> {
+    let iter = vec![1, 2, 3]
+        .into_iter()
+        .map(|x| x + 1);
+    
+    if x % 2 == 0 {
+        Box::new(iter.filter(|x| x % 2 == 0))
+    } else {
+        Box::new(iter)
+    }
+}
+```
+
+Here, we may return a filtered iterator, or maybe not. There's two different
+types that can be returned, and so we must use a trait object.
+
+Oh, and one last thing: to make the syntax a bit more symmetrical, you can
+use `impl Trait` in argument position too. That is:
+
+```rust
+// before
+fn foo<T: Trait>(x: T) {
+
+// after
+fn foo(x: impl Trait) {
+```
+
+which can look a bit nicer for short signatures.
+
+> Side note for you type theorists out there: this isn't an existential, still
+> a universal. In other words, `impl Trait` is universal in an input position, but
+> existential in an output position.
```

Comment:
This section seems to imply `impl Trait` is a replacement for `Box<Trait>`, but it isn't -- `Box<Trait>` is more powerful: it allows for returning both `1.0` (f32) and `2` (i32). I'm not sure it's worth mentioning this in the blog, but maybe a sentence would be good? We may eventually want to extend `impl Trait` to auto-generate impls for an implicit enum, but that's not the case currently.

Reply:
that's covered just above

> It's important to note that sometimes trait objects are still
> what you need. You can only use `impl Trait` if your function returns
> a single type; if you want to return multiple, you need dynamic dispatch.
> For example: